protocols/randhound/proof.go

```diff
@@ -0,0 +1,298 @@
+package randhound
+
+import (
+	"errors"
+
+	"github.com/dedis/cothority/crypto"
+	"github.com/dedis/crypto/abstract"
+	"github.com/dedis/crypto/poly"
+	"github.com/dedis/crypto/random"
+)
+
+// Package (XXX: name) provides functionality to show equality of discrete
+// logarithms (dlog) through non-interactive zero-knowledge (NIZK) proofs.
+
+// Proof resembles a NIZK dlog-equality proof. Allows to handle multiple proofs.
+type Proof struct {
+	suite abstract.Suite
+	base  []ProofBase
+	core  []ProofCore
+}
+
+// ProofBase contains the base points against which the core proof is created.
+type ProofBase struct {
+	g abstract.Point
+	h abstract.Point
+}
+
+// ProofCore contains the core elements of the NIZK dlog-equality proof.
+type ProofCore struct {
+	C  abstract.Scalar // challenge
+	R  abstract.Scalar // response
+	VG abstract.Point  // public commitment with respect to base point G
+	VH abstract.Point  // public commitment with respect to base point H
+}
+
+// NewProof creates a new NIZK dlog-equality proof.
+func NewProof(suite abstract.Suite, g []abstract.Point, h []abstract.Point, core []ProofCore) (*Proof, error) {
+
+	if len(g) != len(h) {
+		return nil, errors.New("Received non-matching number of points")
+	}
+
+	n := len(g)
+	base := make([]ProofBase, n)
+	for i := 0; i < n; i++ {
+		base[i] = ProofBase{g: g[i], h: h[i]}
+	}
+
+	return &Proof{suite: suite, base: base, core: core}, nil
+}
+
+// Setup initializes the proof by randomly selecting a commitment v,
+// determining the challenge c = H(xG,xH,vG,vH), and the response r = v - cx.
+func (p *Proof) Setup(scalar ...abstract.Scalar) ([]abstract.Point, []abstract.Point, []ProofCore, error) {
```

Comment:
I am not sure. I tried to keep it as simple as possible and I am not sure if you get any benefit of having just `[]abstract.Point` in a structure. You would probably have to put all the return values into a structure. I guess that's what you mean? Iirc I tried that a while ago and it complicated quite a few things in the other code that uses PVSS. I'll think about it again.